jsx/Form.js

```diff
@@ -134,6 +134,131 @@ var FormElement = React.createClass({
   }
 });
 
+/**
+ * Datalist Component
+ * React wrapper for a searchable dropdown
+ */
+var DatalistElement = React.createClass({
+
+  propTypes: {
+    name: React.PropTypes.string.isRequired,
+    options: React.PropTypes.object.isRequired,
+    id: React.PropTypes.string,
+    strictDatalist: React.PropTypes.bool,
+    label: React.PropTypes.string,
+    value: React.PropTypes.oneOfType([
+      React.PropTypes.string,
+      React.PropTypes.array
+    ]),
+    class: React.PropTypes.string,
+    disabled: React.PropTypes.bool,
+    required: React.PropTypes.bool,
+    hasError: React.PropTypes.bool,
+    errorMessage: React.PropTypes.string,
+    onUserInput: React.PropTypes.func
+  },
+
+  getDefaultProps: function() {
+    return {
+      name: '',
+      options: {},
+      strictDatalist: false,
+      label: '',
+      value: undefined,
+      id: '',
+      class: '',
+      disabled: false,
+      required: false,
+      hasError: false,
+      errorMessage: 'The field is required!',
+      onUserInput: function() {
+        console.warn('onUserInput() callback is not set');
+      }
+    };
+  },
+
+  getKeyFromValue: function(value) {
+    var options = this.props.options;
+    var keyValue;
+    Object.keys(options).forEach(function(o) {
+      if (options[o] === value) {
+        keyValue = o;
+        return;
+      }
+    }, this);
+
+    return keyValue;
+  },
+
+  handleChange: function(e) {
+    var keyValue = this.getKeyFromValue(e.target.value);
+    this.props.onUserInput(this.props.name, keyValue);
+  },
+
+  handleBlur: function(e) {
+    // null out entry if not present in options in strict mode
+    if (this.props.strictDatalist) {
+      var value = e.target.value;
+      var options = this.props.options;
+      if (Object.values(options).indexOf(value) === -1) {
+        // null out both the hidden value as well as the input text
+        document.querySelector(`input[name="${this.props.name + '_list'}"]`).value = '';
+        this.props.onUserInput(this.props.name, '');
+      }
```

Comment:
You may also want to throw an error message notifying the user that what they have typed is not in the list of possible options, just so that they are not confused. I think I will gain a better understanding of this function upon manual testing.

Reply:
After testing this, I agree with my initial statement. I think an error message stating that what was typed is not an available option would be much clearer for the user. This applies for when the field or required, or not required. I also think that, in the case the field is not required, it may be useful to eliminate the entry entirely from the `formData` rather than set it to an empty string. That's really just a personal preference though, as I rather not have any blank or null values submitted from the `formData` to the ajax call. This, of course, will need to be checked for anyways in the validation of any form.